feat: Standardise LLM prompts with Prompty templates

alice-wondering opened the pull request:
## Purpose

Introduce templates to standardise LLM prompts and support switching between different models at runtime. Added a templates directory to store structured templates for ChatGPT 3.5 and MistralAI-Instruct (from Hugging Face).

Each template consists of the configuration for the model parameters written in YAML, and a Jinja template for prompting models. The backend for Ask and Chat tabs has been modified to use the Prompty objects available in the Promptflow package to load the templates. This feature addresses the issue of incompatible prompting for OpenAI services and the Hugging Face inference API through universalised, structured templates.


## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

- [ ] Yes
- [x] No

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

- [ ] Yes
- [x] No

## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [ ] The current tests all pass (`python -m pytest`).
- [ ] I added tests that prove my fix is effective or that my feature works
- [ ] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [ ] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

alice-wondering committed:
Feat: Templates Introduction

Created templates for hf_mistralai and OpenAI models

Signed-off-by: Alice Cui <alice7301@gmail.com>

alice-wondering committed:
Template backend

Signed-off-by: Alice Cui <alice7301@gmail.com>

nextedoff commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -153,29 +155,39 @@ async def run_until_final_call(
+        current_model = self.hf_model if self.hf_model else self.chatgpt_model
+
+        # Load the Prompty object
+        query_prompty_path = SUPPORTED_MODELS.get(current_model) / "query.prompty"
+        query_prompty = Prompty.load(source=query_prompty_path)
```
The receival of the path from SUPPORTED_MODELS is pretty straightforward, I would just add it as a source instead of separately assigning it to the query_prompty_path
```suggestion
        query_prompty_path = SUPPORTED_MODELS.get(current_model) / "query.prompty"
        query_prompty = Prompty.load(source=SUPPORTED_MODELS.get(current_model) / "query.prompty")
```

nextedoff commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -153,29 +155,39 @@ async def run_until_final_call(
+                ),  # Setting too low risks malformed JSON, setting too high may affect performance
                 n=1,
                 tools=tools,
                 seed=seed,
```
Maybe it wouldn't be a bad idea, as Prompty template can have unlimited LLM parameters that we might not even be aware off, to pass those params from the config as **kwargs, which should probably work.

nextedoff commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -280,9 +292,21 @@ async def run_until_final_call(
+            ),
             n=1,
             stream=should_stream,
             seed=seed,
```
Same as previous, all parameters have to be passed

nextedoff commented on app/backend/approaches/retrievethenread.py:
```diff
@@ -114,18 +117,15 @@ async def run(
-            new_user_content=user_content,
-            max_tokens=self.chatgpt_token_limit - response_token_limit,
+        # Load the Prompty object
+        prompty_path = SUPPORTED_MODELS.get(current_model) / "ask.prompty"
+        prompty = Prompty.load(source=prompty_path)
```
Do what you think, but I would just put it that way ^^
```suggestion
        prompty = Prompty.load(source=SUPPORTED_MODELS.get(current_model) / "ask.prompty")
```

nextedoff commented on app/backend/approaches/retrievethenread.py:
```diff
@@ -135,13 +135,22 @@ async def run(
+                prompty._model.parameters["max_tokens"] if prompty._model.parameters["max_tokens"] is not None else 1024
+            ),
             n=1,
             seed=seed,
```
Same as in 'chatreadretrieveread.py', you need to add other parameters that someone could possibly add in the template, at the moment, it seems like we only support temperature and max_tokens.

nextedoff reviewed (changes requested):
A few changes are needed, the most important one is supporting different parameters than the ones predefined and having default values.
Also, some of the mypy tests failed, you can look into those if you have a minute.

alice-wondering committed:
Fix: PR comments solved

Signed-off-by: Alice Cui <alice7301@gmail.com>

alice-wondering commented:
All comments have been solved apart from the mypy issues, ptal!

nextedoff committed:
Fixed default temperature settings and removed redundant code

Signed-off-by: nextedoff <rajmukx22@gmail.com>

nextedoff commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -273,16 +254,23 @@ async def run_until_final_call(
-            temperature=overrides.get("temperature", 0.3),
-            max_tokens=response_token_limit,
+            messages=ast.literal_eval(messages),
+            **query_prompty._model.parameters,
```
Why does it use query_prompty?

nextedoff committed:
Major bug fixes

In this commit major bugs are fixed that were introduced in the Prompty Templates integration, including inability to answer follow up questions and app not getting the correct template in chat tab

Signed-off-by: Raimund Matulevic <rajmukx22@gmail.com>

phoevos committed:
Merge remote-tracking branch 'origin/main' into feature-template

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

nextedoff committed:
Implemented messages truncation if they are too long

Signed-off-by: nextedoff <rajmukx22@gmail.com>

nextedoff committed:
Updated test snapshots due to template changes

Signed-off-by: nextedoff <rajmukx22@gmail.com>

nextedoff committed:
Clean up

Signed-off-by: nextedoff <rajmukx22@gmail.com>

nextedoff committed:
Bug fix

Signed-off-by: nextedoff <rajmukx22@gmail.com>

phoevos commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -127,61 +129,60 @@ async def run_until_final_call(
+            few_shots=self.query_prompt_few_shots,
+            past_messages=past_messages,
+        )
+        query_messages = ast.literal_eval(query_messages)
```
Why are we evaluating the parsed messages? I understand we get a `str` representation of a list of messages. Is this just to convert it back to a Python `list`?

nextedoff replied:
> Why are we evaluating the parsed messages? I understand we get a `str` representation of a list of messages. Is this just to convert it back to a Python `list`?

Yep, that is solely to convert to a dictionary as for some reason Prompty.render returns a string.

phoevos commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -127,61 +129,60 @@ async def run_until_final_call(
-                        "required": ["search_query"],
-                    },
-                },
-            }
-        ]
+        # Process results
```
I see that we've moved these tools inside the templates. What are they? What are they used for?

nextedoff replied:
From my understanding, tools are not that important as the larger LLMs can do really well with text, though the tool should 'kind of' help with providing consistent results, in this case extracting and creating a query that would be suitable for querying Azure Search.

phoevos commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -273,16 +271,24 @@ async def run_until_final_call(
             ],
         }
+
+        # If the temperature is overridden via the API request, use that value.
+        # Otherwise, use the default value from the model configuration.
+        # If model configuration does not have a temperature, use the default value of 0.3.
+        if overrides.get("temperature") is not None:
+            chat_prompty._model.parameters["temperature"] = overrides.get("temperature")
+        else:
+            chat_prompty._model.parameters.setdefault("temperature", 0.3)
```
Is the temperature the only value that we allow overriding dynamically?

nextedoff replied:
More or less yes, however you are right, there should be a dynamic way to support the overrides.

nextedoff replied:
Comment for myself: 
done

phoevos commented on app/backend/core/message_truncater/__init__.py:
Gotta say, I'm hating the name of the package :laughing:

phoevos commented on app/backend/core/message_truncater/images_helper.py:
Where does this come from? Did we support images before? Do we now (both for OpenAI and arbitrary models)?

nextedoff replied:
We do not, however, it was supported in the original repository responsible for truncating and building the messages. It might be reasonable to use it, if we move the vision approaches to templates too, which I was actually thinking about doing, as templates can be easily added now and we would have consistency across the approaches.

phoevos commented on app/backend/core/message_truncater/images_helper.py:
```diff
@@ -0,0 +1,64 @@
+
+from PIL import Image
+
+"""
+This function checks whether the base code encoded image is a valid base64 string
+and returns the dimensions of the image.
+"""
+
+
+def get_image_dims(image_uri: str) -> tuple[int, int]:
```
Let's make this a proper docstring (underneath the function signature, and with proper formatting).

phoevos commented on app/backend/core/message_truncater/images_helper.py:
```diff
@@ -0,0 +1,64 @@
+"""
+
+
+def count_tokens_for_image(image_uri: str, detail: str = "auto") -> int:
```
Let's make this a proper docstring (underneath the function signature, and with proper formatting).

phoevos commented on app/backend/core/message_truncater/message_shortener.py:
```diff
@@ -0,0 +1,100 @@
+from .model_helper import count_tokens_for_message, count_tokens_for_system_and_tools
+
+
+# Normalises the content of a message to NFC form to ensure the consistency
+# between the messages format.
+def normalize_content(content: Union[str, Iterable[ChatCompletionContentPartParam]]):
```
Let's make this a proper docstring (underneath the function signature, and with proper formatting).

phoevos commented on app/backend/core/message_truncater/message_shortener.py:
```diff
@@ -0,0 +1,100 @@
+    ChatCompletionUserMessageParam,
+)
+
+from .model_helper import count_tokens_for_message, count_tokens_for_system_and_tools
```
Let's use absolute paths again.

phoevos commented on app/backend/core/message_truncater/message_shortener.py:
```diff
@@ -0,0 +1,100 @@
+    past_messages: list[ChatCompletionMessageParam] = [],  # *not* including system prompt
+    fallback_to_default: bool = False,
+) -> list[ChatCompletionMessageParam]:
+    """
```
Let's maintain consistent docstring formatting everywhere.

phoevos reviewed (changes requested):
There's a lot of new code and it's hard for me to dive deeper without the full context, let's talk about this tomorrow! For now I've left a couple of questions and stylistic comments, but this is far from an exhaustive review.

The main thing I'd like to suggest is dropping the `HUGGINGFACE_MODEL` envvar completely. Given that we're moving to supporting multiple models and switching between them at runtime, it makes little sense for this to exist. We can keep an envvar around to allow specifying the "default" model, the one that will be used if a model is not explicitly specified in the request.

Oh yeah, and the message truncater package will absolutely need to be renamed :rofl: I'm not quite sure why all of these helpers are needed, but it seems to me that they could all be under sth like `core/utils` or message utils I guess.

Apart from the above, noting here that we'll have to extend the UI to expose selecting the model from a dropdown list of the supported ones!

nextedoff committed:
Bug fixes and improvements

Signed-off-by: nextedoff <rajmukx22@gmail.com>

phoevos commented on app/backend/templates/hf_mistralai/chat.prompty:
```diff
@@ -0,0 +1,51 @@
+  system_message: Below is a history of the conversation so far, and a new question asked by the user that needs to be answered by searching in a knowledge base.
+  question: What is included in my Northwind Health Plus plan that is not in standard?
+  sources: ["source1.txt: This is data from source1\nsource2.txt:This is data from source 2"]
+  past_messages: [{"role": "user", "content": "What is the weather today"}, {"role": "assistant", "content": "It is sunny today"}]
```
Is this intentionally completely irrelevant?

nextedoff replied:
It is only included to show the format of sample messages, otherwise yes, I think it is irrelevant.
Adding samples that would be tailored to the context of our app would be better, but it is basically to show what is required for the template and in what format.

phoevos commented on app/backend/templates/hf_mistralai/query.prompty:
```diff
@@ -0,0 +1,52 @@
+  system_message: Below is a history of the conversation so far, and a new question asked by the user that needs to be answered by searching in a knowledge base.
+  question: What is included in my Northwind Health Plus plan that is not in standard?
+  few_shots: [{"role": "user", "content": "What is the weather today"}, {"role": "assistant", "content": "It is sunny today"}]
+  past_messages: [{"role": "user", "content": "What is the weather today"}, {"role": "assistant", "content": "It is sunny today"}]
```
Same as above.

phoevos commented on app/backend/templates/openai/chat.prompty:
```diff
@@ -0,0 +1,46 @@
+  system_message: Below is a history of the conversation so far, and a new question asked by the user that needs to be answered by searching in a knowledge base.
+  question: What is included in my Northwind Health Plus plan that is not in standard?
+  sources: ["source1.txt: This is data from source1\nsource2.txt:This is data from source 2"]
+  past_messages: [{"role": "user", "content": "What is the weather today"}, {"role": "assistant", "content": "It is sunny today"}]
```
Same.

phoevos commented on app/backend/templates/openai/query.prompty:
```diff
@@ -0,0 +1,49 @@
+  system_message: Below is a history of the conversation so far, and a new question asked by the user that needs to be answered by searching in a knowledge base.
+  question: What is included in my Northwind Health Plus plan that is not in standard?
+  few_shots: [{"role": "user", "content": "What is the weather today"}, {"role": "assistant", "content": "It is sunny today"}]
+  past_messages: [{"role": "user", "content": "What is the weather today"}, {"role": "assistant", "content": "It is sunny today"}]
```
Same.

phoevos committed:
fix: Drop fastapi from requirements.in

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Update docstrings to match Google style guide

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Drop unused internal function argument

Remove the 'model' argument used by the internal function
'hugging_face_encoding'.

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
Revert "fix: Drop unused internal function argument"

Fucking stupid piece of shit.

The function very obviously needs an argument. I removed `model` because
the linter correctly detects it as unused. We could have an underscore
there but cba, really.

This reverts commit 1c35f4920b5fbc2b558f0437f3da68bbabe6ec71.

phoevos committed:
Merge branch 'main' into feature-template

phoevos reviewed (approved):
LGTM, cheers!

nextedoff reviewed (approved):
LGTM
